addons/account/models/account_account.py

```diff
@@ -589,16 +589,23 @@ def _adapt_accounts_for_account_groups(self, account_ids=None):
         """
         if not self and not account_ids:
             return
-        self.env['account.group'].flush()
-        self.env['account.account'].flush()
+        self.env['account.group'].flush(self.env['account.group']._fields)
+        self.env['account.account'].flush(self.env['account.account']._fields)
         query = """
-            UPDATE account_account account SET group_id = (
-                SELECT agroup.id FROM account_group agroup
-                WHERE agroup.code_prefix_start <= LEFT(account.code, char_length(agroup.code_prefix_start))
-                AND agroup.code_prefix_end >= LEFT(account.code, char_length(agroup.code_prefix_end))
-                AND agroup.company_id = account.company_id
-                ORDER BY char_length(agroup.code_prefix_start) DESC LIMIT 1
-            ) WHERE account.company_id in %(company_ids)s {where_account};
+            WITH relation AS (
+       SELECT DISTINCT FIRST_VALUE(agroup.id) OVER (PARTITION BY account.id ORDER BY char_length(agroup.code_prefix_start) DESC) AS group_id,
```

Comment:
```suggestion
       SELECT DISTINCT FIRST_VALUE(agroup.id) OVER (PARTITION BY account.id ORDER BY char_length(agroup.code_prefix_start) DESC, agroup.id) AS group_id,
``` as I don't know if char_length(agroup.code_prefix_start) is unique per account